File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/statements/expressions/BinaryOperator.java
/*
 * Copyright (c) 2020, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.statements.expressions;

import de.fraunhofer.aisec.cpg.graph.*;
import de.fraunhofer.aisec.cpg.graph.HasType.TypeListener;
import de.fraunhofer.aisec.cpg.graph.types.Type;
import de.fraunhofer.aisec.cpg.graph.types.TypeParser;
import java.util.*;
import org.apache.commons.lang3.builder.ToStringBuilder;
import org.jetbrains.annotations.Nullable;
import org.neo4j.ogm.annotation.Transient;

/**
 * A binary operation expression, such as "a + b". It consists of a left hand expression (lhs), a
 * right hand expression (rhs) and an operatorCode.
 */
public class BinaryOperator extends Expression implements TypeListener, Assignment {

  /** The left hand expression. */
  @SubGraph("AST")
  private Expression lhs;

  /** The right hand expression. */
  @SubGraph("AST")
  private Expression rhs;

  /** The operator code. */
  private String operatorCode;

  /** Required for compound BinaryOperators. This should not be stored in the graph */
  @Transient
  public static final List<String> compoundOperators =
      List.of("*=", "/=", "%=", "+=", "-=", "<<=", ">>=", "&=", "^=", "|=");

  public Expression getLhs() {
    return lhs;
  }

  public <T extends Expression> T getLhsAs(Class<T> clazz) {
    return clazz.isInstance(this.lhs) ? clazz.cast(this.lhs) : null;
  }

  public void setLhs(Expression lhs) {
    if (this.lhs != null) {
      disconnectOldLhs();
    }
    this.lhs = lhs;
    if (lhs != null) {
      connectNewLhs(lhs);
    }
  }

  private void connectNewLhs(Expression lhs) {
    lhs.registerTypeListener(this);
    if ("=".equals(operatorCode)) {
      if (lhs instanceof DeclaredReferenceExpression) {
        // declared reference expr is the left hand side of an assignment -> writing to the var
        ((DeclaredReferenceExpression) lhs).setAccess(AccessValues.WRITE);
      } else if (lhs instanceof MemberExpression) {
        ((MemberExpression) lhs).setAccess(AccessValues.WRITE);
      }
      if (lhs instanceof TypeListener) {
        this.registerTypeListener((TypeListener) lhs);
        this.registerTypeListener((TypeListener) this.lhs);
      }
    } else if (compoundOperators.contains(operatorCode)) {
      if (lhs instanceof DeclaredReferenceExpression) {
        // declared reference expr is the left hand side of an assignment -> writing to the var
        ((DeclaredReferenceExpression) lhs).setAccess(AccessValues.READWRITE);
      } else if (lhs instanceof MemberExpression) {
        ((MemberExpression) lhs).setAccess(AccessValues.READWRITE);
      }
      if (lhs instanceof TypeListener) {
        this.registerTypeListener((TypeListener) lhs);
        this.registerTypeListener((TypeListener) this.lhs);
      }
    }
  }

  private void disconnectOldLhs() {
    this.lhs.unregisterTypeListener(this);
    if ("=".equals(operatorCode) && this.lhs instanceof TypeListener) {
      unregisterTypeListener((TypeListener) this.lhs);
    }
  }

  public Expression getRhs() {
    return rhs;
  }

  public <T extends Expression> T getRhsAs(Class<T> clazz) {
    return clazz.isInstance(this.rhs) ? clazz.cast(this.rhs) : null;
  }

  public void setRhs(Expression rhs) {
    if (this.rhs != null) {
      disconnectOldRhs();
    }
    this.rhs = rhs;
    if (rhs != null) {
      connectNewRhs(rhs);
    }
  }

  private void connectNewRhs(Expression rhs) {
    rhs.registerTypeListener(this);
    if ("=".equals(operatorCode) && rhs instanceof TypeListener) {
      this.registerTypeListener((TypeListener) rhs);
    }
  }

  private void disconnectOldRhs() {
    this.rhs.unregisterTypeListener(this);
    if ("=".equals(operatorCode) && this.rhs instanceof TypeListener) {
      unregisterTypeListener((TypeListener) this.rhs);
    }
  }

  public String getOperatorCode() {
    return operatorCode;
  }

  public void setOperatorCode(String operatorCode) {
    this.operatorCode = operatorCode;
  }

  @Override
  public void typeChanged(HasType src, List<HasType> root, Type oldType) {
    if (!TypeManager.isTypeSystemActive()) {
      return;
    }
    Type previous = this.type;
    if (this.operatorCode.equals("=")) {
      setType(src.getPropagationType(), root);
    } else if (this.lhs != null && "java.lang.String".equals(this.lhs.getType().toString())
        || this.rhs != null && "java.lang.String".equals(this.rhs.getType().toString())) {
      // String + any other type results in a String
      getPossibleSubTypes().clear(); // TODO: Why do we clear the list here?
      setType(TypeParser.createFrom("java.lang.String", getLanguage()), root);
    } else {
      Type resultingType = getLanguage().propagateTypeOfBinaryOperation(this);
      if (resultingType != null) {
        setType(resultingType, root);
      }
    }

    if (!previous.equals(this.type)) {
      this.type.setTypeOrigin(Type.Origin.DATAFLOW);
    }
  }

  @Override
  public void possibleSubTypesChanged(HasType src, List<HasType> root) {
    if (!TypeManager.isTypeSystemActive()) {
      return;
    }
    List<Type> subTypes = new ArrayList<>(getPossibleSubTypes());
    subTypes.addAll(src.getPossibleSubTypes());
    setPossibleSubTypes(subTypes, root);
  }

  @Override
  public String toString() {
    return new ToStringBuilder(this, Node.TO_STRING_STYLE)
        .append("lhs", (lhs == null ? "null" : lhs.getName()))
        .append("rhs", (rhs == null ? "null" : rhs.getName()))
        .append("operatorCode", operatorCode)
        .toString();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (!(o instanceof BinaryOperator)) {
      return false;
    }
    BinaryOperator that = (BinaryOperator) o;
    return super.equals(that)
        && Objects.equals(lhs, that.lhs)
        && Objects.equals(rhs, that.rhs)
        && Objects.equals(operatorCode, that.operatorCode);
  }

  @Override
  public int hashCode() {
    return super.hashCode();
  }

  @Nullable
  @Override
  public AssignmentTarget getTarget() {
    // We only want to supply a target if this is an assignment
    return isAssignment()
        ? (lhs instanceof AssignmentTarget ? (AssignmentTarget) lhs : null)
        : null;
  }

  @Nullable
  @Override
  public Expression getValue() {
    return isAssignment() ? rhs : null;
  }

  public boolean isAssignment() {
    // TODO(oxisto): We need to discuss, if the other operators are also assignments and if we
    // really want them
    return this.operatorCode.equals("=")
    /*||this.operatorCode.equals("+=") ||this.operatorCode.equals("-=")
    ||this.operatorCode.equals("/=")  ||this.operatorCode.equals("*=")*/ ;
  }
}

File: cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/expressions/BinaryOperator.kt
/*
 * Copyright (c) 2020, Fraunhofer AISEC. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *                    $$$$$$\  $$$$$$$\   $$$$$$\
 *                   $$  __$$\ $$  __$$\ $$  __$$\
 *                   $$ /  \__|$$ |  $$ |$$ /  \__|
 *                   $$ |      $$$$$$$  |$$ |$$$$\
 *                   $$ |      $$  ____/ $$ |\_$$ |
 *                   $$ |  $$\ $$ |      $$ |  $$ |
 *                   \$$$$$   |$$ |      \$$$$$   |
 *                    \______/ \__|       \______/
 *
 */
package de.fraunhofer.aisec.cpg.graph.statements.expressions

import de.fraunhofer.aisec.cpg.frontends.TranslationException
import de.fraunhofer.aisec.cpg.graph.ArgumentHolder
import de.fraunhofer.aisec.cpg.graph.HasOverloadedOperation
import de.fraunhofer.aisec.cpg.graph.Node
import de.fraunhofer.aisec.cpg.graph.edges.ast.astEdgeOf
import de.fraunhofer.aisec.cpg.graph.edges.unwrapping
import de.fraunhofer.aisec.cpg.graph.types.HasType
import de.fraunhofer.aisec.cpg.graph.types.Type
import java.util.*
import org.apache.commons.lang3.builder.ToStringBuilder
import org.neo4j.ogm.annotation.Relationship

/**
 * A binary operation expression, such as "a + b". It consists of a left hand expression (lhs), a
 * right hand expression (rhs) and an operatorCode.
 *
 * Note: For assignments, i.e., using an `=` or `+=`, etc. the [AssignExpression] MUST be used.
 */
open class BinaryOperator :
    Expression(), HasOverloadedOperation, ArgumentHolder, HasType.TypeObserver {

    /** The left-hand expression. */
    @Relationship("LHS")
    var lhsEdge =
        astEdgeOf<Expression>(
            of = ProblemExpression("could not parse lhs"),
            onChanged = ::exchangeTypeObserverWithAccessPropagation,
        )
    var lhs by unwrapping(BinaryOperator::lhsEdge)

    /** The right-hand expression. */
    @Relationship("RHS")
    var rhsEdge =
        astEdgeOf<Expression>(
            of = ProblemExpression("could not parse rhs"),
            onChanged = ::exchangeTypeObserverWithAccessPropagation,
        )
    var rhs by unwrapping(BinaryOperator::rhsEdge)

    /** The operator code. */
    override var operatorCode: String? = null
        set(value) {
            field = value
            if (
                (operatorCode in language.compoundAssignmentOperators) ||
                    (operatorCode in language.simpleAssignmentOperators)
            ) {
                throw TranslationException(
                    "Creating a BinaryOperator with an assignment operator code is not allowed. The class AssignExpression must be used instead."
                )
            }
        }

    override fun toString(): String {
        return ToStringBuilder(this, TO_STRING_STYLE)
            .append("lhs", lhs.name)
            .append("rhs", rhs.name)
            .append("operatorCode", operatorCode)
            .append("location", location)
            .toString()
    }

    override fun typeChanged(newType: Type, src: HasType) {
        // We need to do some special dealings for function pointer calls
        if (operatorCode == ".*" || operatorCode == "->*" && src === rhs) {
            // Propagate the function pointer type to the expression itself. This helps us later in
            // the call resolver, when trying to determine, whether this is a regular call or a
            // function pointer call.
            this.type = newType
        } else {
            // Otherwise, we have a special language-specific function to deal with type propagation
            val type =
                language.propagateTypeOfBinaryOperation(this.operatorCode, lhs.type, rhs.type, this)
            this.type = type
        }
    }

    override fun assignedTypeChanged(assignedTypes: Set<Type>, src: HasType) {
        // TODO: replicate something similar like propagateTypeOfBinaryOperation for assigned types
    }

    /** The binary operator operators on the [lhs]. [rhs] is part of the [operatorArguments]. */
    override val operatorArguments: List<Expression>
        get() = listOf(rhs)

    /** The binary operator operators on the [lhs]. [rhs] is part of the [operatorArguments]. */
    override val operatorBase: Expression
        get() = lhs

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other !is BinaryOperator) {
            return false
        }
        return super.equals(other) &&
            lhs == other.lhs &&
            rhs == other.rhs &&
            operatorCode == other.operatorCode
    }

    override fun hashCode() = Objects.hash(super.hashCode(), lhs, rhs, operatorCode)

    override fun addArgument(expression: Expression) {
        if (lhs is ProblemExpression) {
            lhs = expression
        } else {
            rhs = expression
        }
    }

    override fun replaceArgument(old: Expression, new: Expression): Boolean {
        return if (lhs == old) {
            lhs = new
            true
        } else if (rhs == old) {
            rhs = new
            true
        } else {
            false
        }
    }

    override fun hasArgument(expression: Expression): Boolean {
        return lhs == expression || rhs == expression
    }

    override fun getStartingPrevEOG(): Collection<Node> {
        return this.lhs.getStartingPrevEOG()
    }

    val base: Expression?
        get() {
            return if (operatorCode == ".*" || operatorCode == "->*") {
                lhs
            } else {
                null
            }
        }
}
